lib/segment/src/types.rs

```diff
@@ -1292,38 +1298,128 @@ pub struct GeoRadius {
 }
 
 impl GeoRadius {
-    pub fn check_point(&self, lon: f64, lat: f64) -> bool {
+    pub fn check_point(&self, point: &GeoPoint) -> bool {
         let query_center = Point::new(self.center.lon, self.center.lat);
-        query_center.haversine_distance(&Point::new(lon, lat)) < self.radius
+        query_center.haversine_distance(&Point::new(point.lon, point.lat)) < self.radius
     }
 }
 
+#[derive(Deserialize)]
+pub struct GeoPolygonShadow {
+    pub exterior: GeoLineString,
+    pub interiors: Option<Vec<GeoLineString>>,
+}
+
 /// Geo filter request
 ///
-/// Matches coordinates inside the polygon, defined by the given coordinates in order.
+/// Matches coordinates inside the polygon, defined by exterior and interiors.
 #[derive(Debug, Deserialize, Serialize, JsonSchema, Clone, PartialEq)]
+#[serde(try_from = "GeoPolygonShadow")]
 #[serde(rename_all = "snake_case")]
 pub struct GeoPolygon {
-    /// Ordered list of coordinates representing the vertices of a polygon.
-    pub points: Vec<GeoPoint>,
+    pub exterior: GeoLineString,
+    pub interiors: Vec<GeoLineString>,
+}
+
+pub struct GeoPolygonValidationError {
```

Comment:
Instead of custom `GeoPolygonValidationError` I propose to use `OperationError` defined in `segment`crate. If no good variant exists in `OperationError` enum, you can add a new variant. `OperationError` has all conventions and unique formatting format. And instead of `Result<_, OperationError>` you can use `OperationResult<_>`. What do you think about this error handling @agourlay ?

Reply:
I think Ivan makes a good point :+1: 

We already have an existing infrastructure for error, it would be nice to use it instead of this ad-hoc representation.

@zzzz-vincent this should be the very last change!

Reply:
Sure, have reflected this in current change. There are also some other struct using customized Error such as `PointsSelectorValidationError` and `GeoPointValidationError`. We should have a seperate PR to refactor them.